tools/driver/emit_pcm_main.cpp

```diff
@@ -0,0 +1,212 @@
+    using namespace options;
+
+    // Parse frontend command line options using Swift's option table.
+    std::unique_ptr<llvm::opt::OptTable> Table = createSwiftOptTable();
+    unsigned MissingIndex;
+    unsigned MissingCount;
+    llvm::opt::InputArgList ParsedArgs =
+      Table->ParseArgs(Args, MissingIndex, MissingCount, EmitPCMOption);
+    if (MissingCount) {
+      Diags.diagnose(SourceLoc(), diag::error_missing_arg_value,
+                     ParsedArgs.getArgString(MissingIndex), MissingCount);
+      return true;
+    }
+    if (ParsedArgs.hasArg(OPT_UNKNOWN)) {
+      for (const Arg *A : ParsedArgs.filtered(OPT_UNKNOWN)) {
+        Diags.diagnose(SourceLoc(), diag::error_unknown_arg,
+                       A->getAsString(ParsedArgs));
+      }
+      return true;
+    }
+
+    if (ParsedArgs.getLastArg(OPT_help)) {
+      std::string Usage = llvm::sys::path::stem(MainExecutablePath);
+      Usage += " -emit-pcm";
+      Table->PrintHelp(llvm::outs(), Usage.c_str(),
+                       "Precompile explicit Clang modules for Swift",
+                       EmitPCMOption, 0, /*ShowAllAliases*/false);
+      return true;
+    }
+
+    if (const Arg *A = ParsedArgs.getLastArg(options::OPT_module_name)) {
+      ModuleName = A->getValue();
+      // TODO: Check that the module name is a valid identifier and not "Swift".
+    } else {
+      Diags.diagnose(SourceLoc(), diag::error_no_module_name_given);
+      return true;
+    }
+
+    if (const Arg *A = ParsedArgs.getLastArg(options::OPT_sdk))
+      SDKPath = A->getValue();
+
+    if (const Arg *A = ParsedArgs.getLastArg(options::OPT_resource_dir))
+      RuntimeResourcePath = A->getValue();
+
+    if (const Arg *A = ParsedArgs.getLastArg(options::OPT_target))
+      Target = llvm::Triple(llvm::Triple::normalize(A->getValue()));
+    else
+      Target = llvm::Triple(llvm::sys::getDefaultTargetTriple());
+
+    for (const Arg *A : ParsedArgs.filtered(OPT_Xcc)) {
+      ClangImporterOpts.ExtraArgs.push_back(A->getValue());
+    }
+
+    std::vector<std::string> InputFilenames;
+    for (const Arg *A : ParsedArgs.filtered(OPT_INPUT)) {
+      InputFilenames.push_back(A->getValue());
+    }
+    if (InputFilenames.empty()) {
+      Diags.diagnose(SourceLoc(), diag::error_mode_requires_an_input_file);
+      return true;
+    }
+    if (InputFilenames.size() > 1) {
+      Diags.diagnose(SourceLoc(), diag::error_mode_requires_one_input_file);
+      return true;
+    }
+    ModuleMapFilename = InputFilenames[0];
+
+    if (const Arg *A = ParsedArgs.getLastArg(OPT_o)) {
+      OutputFilename = A->getValue();
+    } else {
+      // Default to `<module name>.pcm`.
+      llvm::SmallString<256> Path{ModuleName.c_str()};
+      llvm::sys::path::replace_extension(
+          Path, file_types::getExtension(file_types::TY_ClangModuleFile));
+      OutputFilename = Path.str();
+    }
+
+    return false;
+  }
+};
+
+int emit_pcm_main(ArrayRef<const char *> Args, const char *Argv0,
```

Comment:
There is a large amount of code to effectively make this its own driver, which I'm concerned will get out of sync with the way we import the resulting Clang modules in subtle ways. Could we instead handle this the way the emission of the precompiled bridging header is handled, as a frontend command that relies on all of the same command-line option parsing and setup as a compile job?

Reply:
Making it similar to `-emit-pch` makes sense as a frontend job, especially if you want the Swift driver to at some point in the future be able to construct explicit PCM jobs as part of a compilation, which is a use case I hadn't considered.

For my use case (integration with Bazel), I want to invoke this completely independently, so I'd want it to be a stable driver invocation as well; for example, `-emit-pch` is a frontend-only flag because that job is only created under the hood when `-enable-bridging-pch` is passed to the driver, so I wouldn't want Bazel's integration to rely on invoking the frontend directly since that's considered unstable. As a standalone invocation, I'd want `swiftc -emit-pcm -module-name Foo path/to/some/module.modulemap` to construct only the PCM job and ensure that only modulemap files are supplied as inputs.

Are there any frontend actions other than compilation that I can look at for guidance to model this?

Reply:
I ended up emulating what was done in FrontendTool for `-emit-pch`, but also made the flag available directly to the driver. So now if `-emit-pcm` (or `-dump-pcm`) is used, it's treated as a `SingleCompile` action type, and FrontendTool checks that flag and hands off control to ClangImporter.

I thought I would also need to create a special `FrontendAction` subclass like `GeneratePCHAction`, but that ended up not being necessary—it looks like that would only be needed if the driver wanted to register PCM actions separately as part of a larger compilation job?

Please let me know if this updated version is closer to what you had in mind!

Reply:
I agree that having `swiftc -emit-pcm -module-name Foo path/to/some/module.modulemap` effectively create only a PCM-emitting job makes sense. The "compilation" path is really the right one for the driver here, because you want a lot of the same code to pass through (VFS, output file maps, most compilation options, etc.).